src/Uno.Toolkit.UI/Controls/ZoomContentControl/ZoomContentControl.cs

```diff
@@ -0,0 +1,467 @@
+using Microsoft.UI.Xaml.Data;
+using Microsoft.UI.Xaml.Input;
+using Microsoft.UI.Xaml.Media;
+using Microsoft.UI.Xaml.Navigation;
+using Microsoft.UI.Input;
+#else
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Controls.Primitives;
+using Windows.UI.Xaml.Data;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Navigation;
+using Windows.UI.Input;
+using Windows.Devices.Input;
+#endif
+
+using static Uno.Toolkit.UI.BoundsVisibilityFlag;
+
+namespace Uno.Toolkit.UI;
+
+[TemplatePart(Name = TemplateParts.RootGrid, Type = typeof(Grid))]
+[TemplatePart(Name = TemplateParts.Presenter, Type = typeof(ContentPresenter))]
+[TemplatePart(Name = TemplateParts.VerticalScrollBar, Type = typeof(ScrollBar))]
+[TemplatePart(Name = TemplateParts.HorizontalScrollBar, Type = typeof(ScrollBar))]
+public partial class ZoomContentControl : ContentControl
+{
+	private static class TemplateParts
+	{
+		public const string RootGrid = "PART_RootGrid";
+		public const string Presenter = "PART_Presenter";
+		public const string HorizontalScrollBar = "PART_ScrollH";
+		public const string VerticalScrollBar = "PART_ScrollV";
+	}
+
+	// Fields
+	private ContentPresenter? _presenter;
+	private ScrollBar? _scrollV;
+	private ScrollBar? _scrollH;
+	private Point _lastPosition = new Point(0, 0);
+	private (bool Horizontal, bool Vertical) _movementDirection = (false, false);
+	private bool IsAllowedToWork => (IsEnabled && IsActive && _presenter is not null);
+	private double _previousVerticalScrollValue = double.MinValue;
+	private double _previousHorizontalScrollValue = double.MinValue;
+	private uint _capturedPointerId;
+	private Point _referencePosition;
+
+	// Constructor
+	public ZoomContentControl()
+	{
+		DefaultStyleKey = typeof(ZoomContentControl);
+		Loaded += OnLoaded;
+		SizeChanged += OnSizeChanged;
+	}
+
+	// Events
+	public event EventHandler<EventArgs>? RenderedContentUpdated;
+
+	// Properties
+	private bool ResetZoomAndOffsetWhenInactive { get; set; } = true;
+
+	public Size AvailableSize
+	{
+		get
+		{
+			var vOffset = (AdditionalMargin.Top + AdditionalMargin.Bottom);
+			var hOffset = (AdditionalMargin.Left + AdditionalMargin.Right);
+			return new Size(ActualWidth - hOffset, ActualHeight - vOffset);
+		}
+	}
+
+	// Methods
+	private async Task RaiseRenderedContentUpdated()
+	{
+		await Task.Yield();
+		RenderedContentUpdated?.Invoke(this, EventArgs.Empty);
+	}
+
+	private async void OnVerticalOffsetChanged()
+	{
+		UpdateContentBoundsVisibility();
+		UpdateScrollVisibility();
+		await RaiseRenderedContentUpdated();
+	}
+
+	private void OnHorizontalOffsetChanged()
+	{
+		UpdateContentBoundsVisibility();
+		UpdateScrollVisibility();
+		UpdateHorizontalScrollBarValue();
+	}
+
+	private async void OnZoomLevelChanged()
+	{
+		CoerceZoomLevel();
+		UpdateScrollLimits();
+		UpdateContentBoundsVisibility();
+		UpdateScrollVisibility();
+		await RaiseRenderedContentUpdated();
```

Comment:
Whenever the zoom level gets coerced, I think it will mean that we'll raise `RenderedContentUpdated` twice. That's probably not correct.

Reply:
To be more clear about my thought here:

```
User does ZoomLevel = value > max zoom level.
DP system will call OnZoomLevelChanged
CoerceZoomLevel will set ZoomLevel = value
DP system will call OnZoomLevelChanged again
then we will have that logic run twice. And at least RaiseRenderedContentUpdated has a side effect.
```

*One* of the ways to re-write this is:

```csharp
	private bool CoerceZoomLevel()
	{
		var zoomLevel = ZoomLevel;
		var coercedZoomLevel = Math.Clamp(zoomLevel, MinZoomLevel, MaxZoomLevel);
		if (coercedZoomLevel != zoomLevel)
		{
			ZoomLevel = coercedZoomLevel;
			return true;
		}

		return false;
	}
```

```csharp
	private async void OnZoomLevelChanged()
	{
		if (CoerceZoomLevel())
		{
			// As an actual coerce happened, we will get another OnZoomLevelChanged call.
			// It's important to avoid calling RaiseRenderedContentUpdated twice.
			return;
		}

		UpdateScrollLimits();
		UpdateContentBoundsVisibility();
		UpdateScrollVisibility();
		await RaiseRenderedContentUpdated();
	}
```

That also means, we are missing a test here.